### PR TITLE
Optimise the log formatter

### PR DESCRIPTION
**Pull request by fasaxc**
## Description

Significantly improve the performance and allocation count of our log formatter:

- Remove allocations by using buffer more carefully.
- Use optimised time formatter.
- Pre-compute level+PID strings to avoid call to strings.ToUpper() etc.
- When sorting field keys, avoid allocation by declaring static-sized array.

Benchmarks:
```
Before:

goos: linux
goarch: amd64
pkg: github.com/projectcalico/calico/libcalico-go/lib/logutils
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkLogWithOurFormat-12               	  309297	      3945 ns/op	     874 B/op	      17 allocs/op
BenchmarkLogWithOurFormatFixedFields-12    	  258922	      4549 ns/op	    1242 B/op	      22 allocs/op
BenchmarkLogWithFieldOurFormat-12          	  202881	      5778 ns/op	    1723 B/op	      26 allocs/op
BenchmarkLogWithFieldsOurFormat-12         	  235368	      4985 ns/op	    1741 B/op	      26 allocs/op

After:

goos: linux
goarch: amd64
pkg: github.com/projectcalico/calico/libcalico-go/lib/logutils
cpu: Intel(R) Core(TM) i9-8950HK CPU @ 2.90GHz
BenchmarkLogWithOurFormat-12               	  429015	      2935 ns/op	     745 B/op	       7 allocs/op
BenchmarkLogWithOurFormatFixedFields-12    	  375942	      2896 ns/op	     985 B/op	       7 allocs/op
BenchmarkLogWithFieldOurFormat-12          	  310732	      3530 ns/op	    1434 B/op	      10 allocs/op
BenchmarkLogWithFieldsOurFormat-12         	  290439	      3830 ns/op	    1484 B/op	      11 allocs/op
```

## Todos

- [ ] Tests
- [ ] Documentation
- [ ] Release note

## Release Note

```release-note
Log formatting performance has been improved, reducing the overhead of logging.
```

## Reminder for the reviewer

Make sure that this PR has the correct labels and milestone set.

Every PR needs one `docs-*` label.

- `docs-pr-required`: This change requires a change to the documentation that has not been completed yet.
- `docs-completed`: This change has all necessary documentation completed.
- `docs-not-required`: This change has no user-facing impact and requires no docs.

Every PR needs one `release-note-*` label.

- `release-note-required`: This PR has user-facing changes. Most PRs should have this label.
- `release-note-not-required`: This PR has no user-facing changes.

Other optional labels:

- `cherry-pick-candidate`: This PR should be cherry-picked to an earlier release. For bug fixes only.
- `needs-operator-pr`: This PR is related to install and requires a corresponding change to the operator.

### CODE REVIEW

**Review by aaaaaaaalex — APPROVED** [review]
LGTM

File: libcalico-go/lib/logutils/logutils.go
Comment (fasaxc) [reply libcalico-go/lib/logutils/logutils.go:201]: Nothing here actually, maybe I should revert that tweak.  It came about because I moved code around and ended up back almost where I started!

Comment (fasaxc) [reply libcalico-go/lib/logutils/logutils.go]: Reverted that tweak.

**Comment by marvin-tigera** [comment]
OK, I will merge the pull request when it's ready, leave the commits as is when I merge it, and leave the branch after I've merged it.

**Pull request merged**
